File: 2_C++(VisualStudio)/2_workspace_MFC_Direct2D/1_SideViewGameSample/src/src/GoalObject.cpp
#include "pch.h"
#include "include/GoalObject.h"

GoalObject::GoalObject():
    radius(16.0f)
    , color(D2D1::ColorF::GreenYellow) {

}

GoalObject::GoalObject(float _radius, D2D1::ColorF _color):
    radius(_radius)
    ,color(_color){

}

TileType GoalObject::GetTileType() {

    return TileType::Goal;
}

void GoalObject::OnHit() {

    isDisappearance = true;
}

void GoalObject::Draw(Renderer& renderer, int x, int y, float offsetY, const Camera& cameraX){

    //拡大率(棒人間と合わせること）
    const float bigRate = 1.3f;

    // --- カメラ変換（世界 → 画面） ---
    x = cameraX.WorldToScreenX(x * TILE_SIZE) * bigRate; //カメラのX座標をスクロールに使用
    y = y * TILE_SIZE * bigRate + offsetY;

    //拡大係数
    float coeX = cameraX.GetScaleX();

    // スケールされた寸法
    float r = radius * coeX * bigRate;

    renderer.DrawCircle(x, y, r, color);
}

File: 99_AI生成ソフトウェア(バイブコーディング)/1_Webアプリ系(VSCode)/3_SideViewGameSample_TypeScript_移植元CPlus(Claude)/SideViewGameSample_移植元CPlus/src/GameLogic/GameObject/GoalObject.ts
import { Renderer } from "../Renderer";
import { Camera } from "../Camera";
import { TileType } from "../TileType";
import { IColorF, ColorF } from "../ColorF";
import { TileMapObject } from "./TileMapObject";

/**
 * GoalObject.ts
 * 移植元: include/GoalObject.h / src/GoalObject.cpp
 *
 * ゴール地点(黄緑の円)。接触(OnHit)すると isDisappearance_ が true になる。
 *
 * 【移植元の挙動をそのまま維持(バグ含む)】
 * 1) Draw() のオーバーライド元シグネチャは x, y が int 型のため、
 *    ワールド→スクリーン座標変換後の浮動小数値を代入する際に
 *    暗黙の int 切り捨てが発生していた。TypeScriptには int 型が無いため、
 *    Math.trunc() でこの切り捨て挙動を明示的に再現している。
 * 2) スケール済み半径 r (= radius * coeX * bigRate) が計算されているが、
 *    実際の DrawCircle 呼び出しでは使われておらず、スケールされていない
 *    素の radius が使われている(移植元の未使用計算値バグをそのまま維持)。
 */
export class GoalObject extends TileMapObject {
    private radius: number;
    private color: IColorF;

    public constructor(radius?: number, color?: IColorF) {
        super();
        this.radius = radius !== undefined ? radius : 16.0;
        this.color = color !== undefined ? color : ColorF.GreenYellow;
    }

    public override getTileType(): TileType {
        return TileType.Goal;
    }

    public override onHit(): void {
        this.isDisappearance_ = true;
    }

    public override draw(renderer: Renderer, x: number, y: number, offsetY: number, cameraX: Camera): void {
        const bigRate:number = 1.3;

        // 移植元は x, y が int 型のパラメータのため、以下の代入で暗黙の切り捨てが発生する
        x = Math.trunc(cameraX.worldToScreenX(x * TileMapObject.TILE_SIZE) * bigRate);
        y = Math.trunc(y * TileMapObject.TILE_SIZE * bigRate + offsetY);

        // 拡大係数
        const coeX:number = cameraX.getScaleX();

        // スケールされた半径
        const r:number = this.radius * coeX * bigRate;

        renderer.drawCircle(x, y, r, this.color);
    }
}
